File: app/projection/datastore/projection_datastore_shard.py
from typing import Dict, List

from app.projection.datastore.projection_file_store import ProjectionFileStore


class ProjectionDatastoreShard:
    '''
    A single shard for projection data storage. Corresponds to one file
    storing projection vectors.
    '''

    def __init__(self, file_store: ProjectionFileStore, projection_filepath: str, in_memory: bool):
        self.projection = dict()
        self.file_store = file_store
        self.projection_filepath = projection_filepath
        self.in_memory = in_memory

    def _load_projection(self) -> None:
        if self.file_store.check_if_object_exists(self.projection_filepath):
            self.projection = self.file_store.get_object(self.projection_filepath)

    def _save_data(self, projection: Dict[str, List[float]]) -> None:
        self.file_store.put_object(self.projection_filepath, projection)

    def _cache_data(self, projection: Dict[str, List[float]]) -> None:
        self.projection = projection

    def load_data(self):
        if not self.in_memory:
            self._load_projection()

    def upload(self, projection: Dict[str, List[float]]) -> None:
        if not self.in_memory:
            self._save_data(projection)
        self._cache_data(projection)

    def get_all(self) -> Dict[str, List[float]]:
        return self.projection

    def get(self, author: str):
        return self.projection[author]

File: experimental/steven/cpp/datastores/projection_datastore_shard.cpp
#include "projection_datastore_shard.h"

ProjectionDatastoreShard::ProjectionDatastoreShard(
    ProjectionFileStore& file_store, std::string projection_filepath)
    : file_store(file_store), projection_filepath(projection_filepath) {}

void ProjectionDatastoreShard::upload(
    std::map<std::string, std::vector<float>> projection) {
  saveData(projection);
  cacheData(projection);
}

void ProjectionDatastoreShard::loadData() { this->loadProjection(); }

std::map<std::string, std::vector<float>> ProjectionDatastoreShard::getAll() {
  return projection;
}

std::vector<float> ProjectionDatastoreShard::get(std::string author_name) {
  return this->projection[author_name];
}

void ProjectionDatastoreShard::cacheData(
    std::map<std::string, std::vector<float>> projection) {
  this->projection = projection;
}

void ProjectionDatastoreShard::cacheData(
    std::map<std::string, std::vector<float>> projection) {
  this->projection = projection;
}

void ProjectionDatastoreShard::saveData(
    std::map<std::string, std::vector<float>> projection) {
  this->file_store.putProjection(this->projection_filepath, projection);
}

void ProjectionDatastoreShard::loadProjection() {
  if (this->file_store.objectExists(this->projection_filepath)) {
    this->projection =
        this->file_store.getProjection(this->projection_filepath);
  }
}
